api/dbadapters/bigquery.ts

```diff
@@ -257,54 +248,82 @@ export class BigQueryDbAdapter implements IDbAdapter {
     return persistedMetadata;
   }
 
-  public async persistStateMetadata(actions: dataform.IExecutionAction[]): Promise<void> {
-    if (actions.length === 0) {
+  public async persistStateMetadata(
+    allActions: dataform.IExecutionAction[],
+    actionsToPersist: dataform.IExecutionAction[],
+    options: {
+      onCancel: OnCancel;
+    }
+  ): Promise<void> {
+    if (allActions.length === 0) {
       return;
     }
-    const allInvolvedTargets = new StringifiedSet(JSONObjectStringifier.create<dataform.ITarget>());
-    actions.forEach(action => {
-      allInvolvedTargets.add(action.target);
-      action.transitiveInputs.forEach(transitiveInput => allInvolvedTargets.add(transitiveInput));
-    });
-
-    const tableMetadataByTarget = new StringifiedMap<dataform.ITarget, dataform.ITableMetadata>(
-      JSONObjectStringifier.create()
-    );
-    await Promise.all(
-      Array.from(allInvolvedTargets).map(async target => {
-        tableMetadataByTarget.set(target, await this.table(target));
-      })
-    );
-    const queries = actions.map(action => {
-      const persistTable = dataform.PersistedTableMetadata.create({
-        target: action.target,
-        lastUpdatedMillis: tableMetadataByTarget.get(action.target).lastUpdatedMillis,
-        definitionHash: hashExecutionAction(action),
-        transitiveInputTables: action.transitiveInputs.map(transitiveInput => {
-          if (!tableMetadataByTarget.has(transitiveInput)) {
-            throw new Error(`Could not find table metadata for ${JSON.stringify(transitiveInput)}`);
-          }
-          return tableMetadataByTarget.get(transitiveInput);
-        })
-      });
-
-      const targetName = `${action.target.database}.${action.target.schema}.${action.target.name}`;
-
-      return buildQuery(targetName, persistTable);
-    });
+    const targetStringifier = JSONObjectStringifier.create<dataform.ITarget>();
+    try {
+      // Create the cache table, if needed.
+      await this.execute(
+        `
+CREATE TABLE IF NOT EXISTS \`${CACHED_STATE_TABLE_NAME}\` (
+  target STRING,
+  metadata_proto STRING
+)`,
+        options
+      );
+      // Before saving any new data, delete all entries for 'allActions'.
+      await this.execute(
+        `
+DELETE \`${CACHED_STATE_TABLE_NAME}\` WHERE target IN (${allActions
```

Comment:
It does seem a shame to loose the atomicity of merge, why change it, is it just to simplify?

Reply:
it wasn't really ever atomic, this was always happening, I've just moved the code around. (a) yes IMO it's simpler, (b) we need to separate out `INSERT` queries anyway due to query size limits, (c) IMO it's easier to reason about: everything is deleted *before* anything gets updated, and if the delete fails, nothing is updated.